lib/core/src/user_administration.cpp

```diff
@@ -0,0 +1,483 @@
+            users.emplace_back(row[0], row[1]);
+        }
+
+        return users;
+    }
+
+    auto groups(rxComm& conn) -> std::vector<group>
+    {
+        std::vector<group> groups;
+
+        // Some groups will not show up because of the condition.
+        // (e.g. "public" will not show up via this query).
+        for (auto&& row : irods::query{&conn, "select USER_GROUP_NAME where USER_TYPE = 'rodsgroup'"}) {
+            groups.emplace_back(row[0]);
+        }
+
+        return groups;
+    }
+
+    auto groups(rxComm& conn, const user& user) -> std::vector<group>
+    {
+        std::vector<group> groups;
+
+        std::string gql = "select USER_GROUP_NAME where USER_TYPE = 'rodsgroup' and USER_NAME = '";
+        gql += unique_name(conn, user);
+        gql += "' and USER_ZONE = '";
+        gql += (user.zone.empty() ? get_local_zone(conn) : user.zone);
+        gql += "'";
+
+        for (auto&& row : irods::query{&conn, gql}) {
+            groups.emplace_back(row[0]);
+        }
+
+        return groups;
+    }
+
+    auto exists(rxComm& conn, const user& user) -> bool
+    {
+        std::string gql = "select USER_ID where USER_TYPE != 'rodsgroup' and USER_NAME = '";
+        gql += user.name;
+        gql += "' and USER_ZONE = '";
+        gql += (user.zone.empty() ? get_local_zone(conn) : user.zone);
+        gql += "'";
+
+        for (auto&& row : irods::query{&conn, gql}) {
+            static_cast<void>(row);
+            return true;
+        }
+
+        return false;
+    }
+
+    auto exists(rxComm& conn, const group& group) -> bool
+    {
+        std::string gql = "select USER_GROUP_ID where USER_TYPE = 'rodsgroup' and USER_GROUP_NAME = '";
+        gql += group.name;
+        gql += "'";
+
+        for (auto&& row : irods::query{&conn, gql}) {
+            static_cast<void>(row);
+            return true;
+        }
+
+        return false;
+    }
+
+    auto id(rxComm& conn, const user& user) -> std::optional<std::string>
+    {
+        std::string gql = "select USER_ID where USER_TYPE != 'rodsgroup' and USER_NAME = '";
+        gql += unique_name(conn, user);
+        gql += "' and USER_ZONE = '";
+        gql += (user.zone.empty() ? get_local_zone(conn) : user.zone);
+        gql += "'";
+
+        for (auto&& row : irods::query{&conn, gql}) {
+            return row[0];
+        }
+
+        return std::nullopt;
+    }
+
+    auto id(rxComm& conn, const group& group) -> std::optional<std::string>
+    {
+        std::string gql = "select USER_GROUP_ID where USER_TYPE = 'rodsgroup' and USER_GROUP_NAME = '";
+        gql += group.name;
+        gql += "'";
+
+        for (auto&& row : irods::query{&conn, gql}) {
+            return row[0];
+        }
+
+        return std::nullopt;
+    }
+
+    auto type(rxComm& conn, const user& user) -> std::optional<user_type>
+    {
+        std::string gql = "select USER_TYPE where USER_TYPE != 'rodsgroup' and USER_NAME = '";
```

Comment:
there's 12 instances of 'rodsgroup' repeated in here - should we pull that out to a single point of truth?

Reply:
I don't think so. If we did, the code would have to add extra `+=` operations to form the GenQuery string which we would pay for.

Reply:
fair enough - it's well contained.